### ISSUE TITLE
Incorrectly reported undefined name in a loop

### ISSUE THREAD

**Issue by hotpxl** [astral-sh/ruff#6569]
Running `ruff --check` (no configuration needed) on 
```
def main():
    for i in range(100):
        if i != 0:
            print(a)
        else:
            a = 0
```
yields an undefined name error and a "assigned to but never used" error, whereas
```
def main():
    for i in range(100):
        if i == 0:
            a = 0
        else:
            print(a)
```
is fine.

While one could argue that the former is bad style and hard to read, unfortunately there's quite a bit of code written in that style already. Is this something Ruff could understand?

**Comment by charliermarsh** [comment]
Thanks for filing! Hmm, while I wish we could, I don't think we can reliably support this kind of conditional assignment. Even Pyright which has a much more sophisticated inference model marks `a` as possibly unbound in both cases. Ideally, we'd mark both cases as "possibly unbound", but even that wouldn't necessarily solve your problem, since you'd still have to fight that lint error.
